core/statix.multilang/src/main/java/mb/statix/multilang/utils/MessageUtils.java

```diff
@@ -110,26 +132,41 @@ private static String cleanupString(String string) {
     }
 
     public static @Nullable ResourceKey resourceKeyFromOrigin(ITerm origin) {
-        if(origin.getAttachments().containsKey(TermIndex.class)) {
-            TermIndex termIndex = (TermIndex)origin.getAttachments().get(TermIndex.class);
-            String resource = termIndex.getResource();
-            String[] split = resource.split(QualifiedResourceKeyString.separator);
-            if(split.length < 2) {
-                return new DefaultResourceKey(null, resource);
-            }
-            final String qualifier = split[0];
-            return new DefaultResourceKey(qualifier.isEmpty() ? null : qualifier, split[1]);
+        if(origin.getAttachments().containsKey(TermOrigin.class)) {
+            TermOrigin termOrigin = (TermOrigin)origin.getAttachments().get(TermOrigin.class);
+            ITermAttachment parent = termOrigin.getImploderAttachment();
+            return getResourceKeyFromParentAttachments(parent);
         }
         return null;
     }
 
+    private static ResourceKey getResourceKeyFromParentAttachments(ITermAttachment attachment) {
+        if(attachment == null) {
+            return null;
+        }
+        if(attachment instanceof ResourceKeyAttachment) {
+            return ((ResourceKeyAttachment)attachment).resourceKey;
+        }
+
+        if(attachment instanceof ParentAttachment) {
+            IStrategoTerm parentTerm = ((ParentAttachment)attachment).getParent();
+            ISimpleTerm root = ParentAttachment.getRoot(parentTerm);
```

Comment:
There seems to be some code duplication with ResourceKeyAttachment (getResourceKey) here.

Reply:
Not exactly. This code also traverses the `ImploderAttachment`. I suspect this is needed because the terms this code is called on are transformed to nabl terms, and then back again.